### PR TITLE
Feature/select new data

### PR DESCRIPTION
**Pull request by Qouagga**
Implement dataset selection for importation. DatasetList class components change to function components.

### CODE REVIEW

File: backend/tests/views/dataset.py [hunk 1]
```diff
@@ -92,38 +111,40 @@ def test_datawork_import_for_staff_good(self):
...
         original_permissions = settings.DATASET_IMPORT_FOLDER.stat().st_mode
         settings.DATASET_IMPORT_FOLDER.chmod(0o444)  # removing open-folder permission
         try:
-            response = self.client.get(url)
+            response = response = self.client.post(
+                url, json.dumps(data_send), content_type="application/json", follow=True
+            )
         finally:
             settings.DATASET_IMPORT_FOLDER.chmod(original_permissions)
-        self.assertContains(response, "Permission denied", status_code=400)
```
Comment (ixio) [inline backend/tests/views/dataset.py:117]: Why have you removed status_code checking ? I mean if you changed this to 401 to 403 just change the number not remove the check :wink: (though I guess here you really changed it to 200).

File: backend/api/actions/datawork_import.py [hunk 2]
```diff
@@ -49,13 +56,13 @@ def datawork_import(*, importer):
...
-            start=parse_datetime(audio_raw["start_date"]),
-            end=parse_datetime(audio_raw["end_date"]),
-        )
+            audio_metadatum = AudioMetadatum.objects.create(
+                num_channels=audio_raw["nchannels"],
+                sample_rate_khz=audio_raw["dataset_fs"],
+                sample_bits=audio_raw["sound_sample_size_in_bits"],
+                start=parse_datetime(audio_raw["start_date"]),
+                end=parse_datetime(audio_raw["end_date"]),
+            )
```
Comment (ixio) [inline backend/api/actions/datawork_import.py:65]: Why did you move all of this inside the `open` ? The reason it was outside was so that we could open&close the file as quickly as possible :thinking:

File: backend/api/views/dataset.py [hunk 3]
```diff
@@ -26,30 +30,73 @@ def list(self, request):
...
+        csv_dataset_names = []
+        new_datasets = []
+
+        # Check for new datasets
+        try:
+            with open(
+                settings.DATASET_IMPORT_FOLDER / "datasets.csv", encoding="utf-8"
+            ) as csvfile:
+                for dataset in csv.DictReader(csvfile):
+                    csv_dataset_names.append(dataset["name"])
```
Comment (ixio) [inline backend/api/views/dataset.py:58]: This is not really used, is it ? :sweat_smile:

File: backend/api/actions/datawork_import.py [hunk 4]
```diff
@@ -17,22 +17,29 @@
 
 
 @transaction.atomic
-def datawork_import(*, importer):
+def datawork_import(*, dataset_checked, importer):
```
Comment (ixio) [inline backend/api/actions/datawork_import.py:20]: If I understand correctly this is named like so because in the frontend this corresponds to datasets that have been checked on the list by the user ? Sending this name to the view is fine but here you should really use a meaningful name to the current context: `datasets_to_import` or `wanted_datasets`, something of the kind :wink:

File: backend/api/views/dataset.py [hunk 5]
```diff
@@ -26,30 +30,73 @@ def list(self, request):
...
+                importer=request.user,
+            )
         except FileNotFoundError as error:
             capture_exception(error)
-            return HttpResponse(error, status=400)
+            messages.add_message(request, messages.ERROR, f"File Not Found : {error}")
+            return redirect(reverse("dataset-list"))
         except PermissionError as error:
             capture_exception(error)
-            return HttpResponse(error, status=400)
```
Comment (ixio) [inline backend/api/views/dataset.py:45]: Looking at the tests I'm guessing your new system returns a 200 response ? Why though ? For a REST API a 400 response is much more semantically appropriate and you should be able to manage it in the front same as a 200 :thinking:

File: backend/settings/base.py [hunk 6]
```diff
@@ -37,6 +37,7 @@
     "django.middleware.csrf.CsrfViewMiddleware",
     "django.contrib.auth.middleware.AuthenticationMiddleware",
     "django.contrib.messages.middleware.MessageMiddleware",
+    "django.contrib.sessions.middleware.SessionMiddleware",
```
Comment (ixio) [inline backend/settings/base.py:40]: This is probably fine but can you tell me why it is needed ?

**Review by ixio — COMMENTED** [review]
Okay good work overall but I'm going to need some changes.

Also it's pretty weird, you seem to have rebased but the diff here is like master isn't up to date: maybe I can look this over with you later on today.

File: frontend/src/DatasetList.js [hunk 7]
```diff
@@ -1,119 +1,189 @@
...
+
+    const importNewData = (dataset_checked) => {
+        return startImport.set('Authorization', 'Bearer ' + props.app_token).send({dataset_checked}).set('Accept', 'application/json').then((req) => {
+
+                let newDatasets_tmp = newDatasets;
+                req.body.forEach(oneNewData => {
+                    newDatasets_tmp = newDatasets_tmp.filter((item) => {
+                        return item.name !== oneNewData.name
+                    });
+                })
```
Comment (rvovard) [inline frontend/src/DatasetList.js:136]: I would have done something like:

```js
const remainingDatasets = newDatasets.filter(newDataset =>
  req.body.some(importedDataset =>
    importedDataset.name !== newDataset.name
  )
);
```

(and the indentation is wrong too)

File: frontend/test/DatasetList.test.js [hunk 8]
```diff
@@ -6,35 +6,35 @@ import { mount, shallow } from 'enzyme';
 import DatasetList from '../src/DatasetList';
 import datasets from './fixtures/dataset_list.json';
 
-describe('testing DatasetList component', function () {
-    this.timeout(20000);
+// describe('testing DatasetList component', function () {
+//     this.timeout(20000);
```
Comment (rvovard) [inline frontend/test/DatasetList.test.js:10]: You can rebase your branch on the new `master` now that faulty tests have been quoted out.
Though it seems that you quoted out all the tests for this component. I think we will leave it as it is now (for the sake of efficiency), but it would be nice to repair at least a few of them in the next weeks.

File: frontend/src/modalNewData.js [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+type NewDataItem_type = {
+    data: object,
+    index: integer
+}
+
+const NewDataItem = (props : NewDataItem_type) => {
+    return (
+        <li className='list-group-item checkbox'>
+            <div className="form-check form-check-reverse">
```
Comment (rvovard) [inline frontend/src/modalNewData.js:19]: I **know** I can be irritating with formatting stuff, but can we be consistent with quotes? I would use double-quotes for HTML attribute values.

To be fair: I need to add a linter in next version of the app (as black on Python side).

File: frontend/src/modalNewData.js [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                    </div>
+                </div>
+            </div>
+        </div>
+    </React.Fragment>
+    )
+}
+
+
+const Modal = (props : modalProp) => props.openModal ? ReactDOM.createPortal (
```
Comment (rvovard) [inline frontend/src/modalNewData.js:191]: Just a thought here, no need to change it but: it would have been nice that the `Modal` component was generic.

**Review by rvovard — CHANGES_REQUESTED** [review]
Nice work overall!
Some details could be improved. Please reach me if you think I'm being too strict or just annoying 😄

File: backend/tests/views/dataset.py [hunk 11]
```diff
@@ -1,5 +1,6 @@
 """Dataset DRF-Viewset test file"""
-
+import json
+from cmath import log
```
Comment (ixio) [inline backend/tests/views/dataset.py:3]: Is this used ? I'm surprised I'm not seeing it but I'm surprised Pylint is not throwing a warning about it :thinking:

File: frontend/src/DatasetList.js [hunk 12]
```diff
@@ -1,119 +1,198 @@
...
+      .catch((err) => {
+        if (err.status && err.status === 401) {
+          // Server returned 401 which means token was revoked
+          document.cookie = "token=;max-age=0;path=/";
+          window.location.reload();
+        }
+        setError(err);
+        setOpenModal(false);
+      });
+  };
```
Comment (rvovard) [inline frontend/src/DatasetList.js:142]: As seen with @Qouagga: the body of this function should be in a `useEffect` hook as it's an asynchronous call to the API.
Best way to do it is to create a boolean dedicated `isImportReady` or `launchImport`, pass this boolean to the modal instead of the `importNewData` function, and use it as a followed value by the useEffet hook.
The hook checks the value of the boolean, sets it to `false` immediatly and launches the API.

Comment (rvovard) [reply frontend/src/DatasetList.js:142]: And don't forget to `.abort()` ^^

**Review by rvovard — CHANGES_REQUESTED** [review]
One last thing ;)

**Review by rvovard — APPROVED** [review]
Thanks for all the work!
